{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c25551ef-a84e-47fb-9b62-7f584ad5a002",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Author Kanishk Asthana kasthana@eng.ucsd.edu\n",
    "import pysam\n",
    "from datetime import datetime,timedelta\n",
    "import numpy as np\n",
    "import argparse\n",
    "import os\n",
    "import sys\n",
    "from multiprocessing import Pool, Queue\n",
    "import copy\n",
    "\n",
    "#For writing output to the stdout in realtime\n",
    "def write(*something):\n",
    "    print(*something)\n",
    "    sys.stdout.flush()\n",
    "\n",
    "#Default Values\n",
    "bamFileName=\"\"\n",
    "outBamFileName=\"\"\n",
    "MIN_UMI_COUNT=20 #Minimum UMI count for Barcode to be considered for Collapse. Not counting UMIs for simplicity.Change this later because it breaks easily. #ToDO\n",
    "SUBSTITUTION_ERROR_FREQ=0.85 #Minimum frequency of majority barcode in substitution pairs needed for collapse\n",
    "max_cores=1\n",
    "\n",
    "def parse_file(input_filename):\n",
    "    if not os.path.isfile(input_filename):\n",
    "        raise argparse.ArgumentTypeError(\"File does not exist. Please use a valid file path.\")\n",
    "    return(input_filename)\n",
    "    \n",
    "def check_umi(input_umi):\n",
    "    value=int(input_umi)\n",
    "    assert value>0,\"Please enter positive values only for UMI Count Filter!\"\n",
    "    return value\n",
    "\n",
    "parser=argparse.ArgumentParser(description=\"Script to detect and Correct Substitution errors in the Cell Barcodes. These errors are likely sequencing errors and not Bead Synthesis Errors\")\n",
    "parser.add_argument(\"INPUT_FILENAME\",help=\"Aligned Merged and Gene labeled BAM file for Correcting Substitution Errors.\", type=parse_file)\n",
    "parser.add_argument(\"OUTPUT_FILENAME\",help=\"Please enter a Valid Path for the Error Corrected output BAM file.\",type=str)\n",
    "parser.add_argument(\"NUM_CORE\",help=\"Maximum number of Cores to be used for the Computation.\",type=int)\n",
    "parser.add_argument(\"-umi\",\"--MIN_UMI\", help=\"Minimum UMI count per barcode to be considered for Collapse. Default Value is 20.\",type=check_umi)\n",
    "\n",
    "args=parser.parse_args()\n",
    "write(args)\n",
    "\n",
    "if args.MIN_UMI is not None:\n",
    "    MIN_UMI_COUNT=args.MIN_UMI\n",
    "\n",
    "bamFileName=args.INPUT_FILENAME\n",
    "outBamFileName=args.OUTPUT_FILENAME\n",
    "max_cores=args.NUM_CORE\n",
    "\n",
    "script_start_time=datetime.now()\n",
    "\n",
    "class CellBarcode:\n",
    "    CellBarcodesWithEnoughCounts=[]\n",
    "    NumberofBarcodesCollapsed=0\n",
    "    def __init__(self,barcode,umi):\n",
    "        self.count=1 #Number of reads with that Barcode. When initializing you count the barcode you initialize with\n",
    "        self.barcode=barcode\n",
    "        self.barcodeToReturn=self #This will be updated with a different reference if this barcode is merged with another\n",
    "        self.umi_dict={}\n",
    "        self.umi_dict[umi]=1\n",
    "        \n",
    "    def increase_count(self,umi):\n",
    "        self.count+=1 #Increase count if you see a barcode\n",
    "        if umi in self.umi_dict:\n",
    "            self.umi_dict[umi]+=1\n",
    "        else:\n",
    "            self.umi_dict[umi]=1\n",
    "            \n",
    "    def computeHasEnoughCounts(self):\n",
    "        if len(self.umi_dict.keys())>MIN_UMI_COUNT:\n",
    "            CellBarcode.CellBarcodesWithEnoughCounts.append(self.barcode)\n",
    "            \n",
    "    def combineIfSubstitution(self,CellBarcode2):\n",
    "        largerBarcode=self.barcodeToReturn\n",
    "        smallerBarcode=CellBarcode2.barcodeToReturn\n",
    "        if smallerBarcode.count>largerBarcode.count:\n",
    "            largerBarcode=CellBarcode2.barcodeToReturn\n",
    "            smallerBarcode=self.barcodeToReturn\n",
    "        freq=largerBarcode.count/(largerBarcode.count+smallerBarcode.count)\n",
    "        if freq>SUBSTITUTION_ERROR_FREQ:\n",
    "            smallerBarcode.barcodeToReturn=largerBarcode\n",
    "            largerBarcode.count+=smallerBarcode.count\n",
    "            CellBarcode.NumberofBarcodesCollapsed+=1\n",
    "\n",
    "barcode_dict={}\n",
    "\n",
    "#If you get an error here your file probably not correctly formated. Make sure you have a header.\n",
    "bamFile=pysam.AlignmentFile(bamFileName,\"rb\")\n",
    "BamRecords=bamFile.fetch(until_eof=True)\n",
    "\n",
    "start_time=datetime.now()\n",
    "prevMil=start_time\n",
    "write(\"Started Processing BAM file at\",start_time,\". Getting Cell Barcodes to correct Illumina Sequencing Base Substitution Errors!\")\n",
    "\n",
    "total_records=0\n",
    "for record in BamRecords:\n",
    "    \n",
    "\n",
    "    #For printing progress\n",
    "    total_records+=1\n",
    "    if total_records%1000000==0:\n",
    "        time_taken=datetime.now()-prevMil\n",
    "        write(\"Finished processing \",total_records,\"\\trecords at\",datetime.now(),\". Previous 1000000 Records took \",time_taken.total_seconds(),\"s\")\n",
    "        prevMil=datetime.now()\n",
    "    \n",
    "\n",
    "    #Main Logic\n",
    "    cell_barcode=record.get_tag('XC')\n",
    "    umi=record.get_tag('XM')\n",
    "    \n",
    "    if cell_barcode in barcode_dict:\n",
    "        barcode_dict[cell_barcode].increase_count(umi)\n",
    "    else:\n",
    "        barcode_dict[cell_barcode]=CellBarcode(cell_barcode,umi)\n",
    "\n",
    "\n",
    "total_time=datetime.now()-start_time\n",
    "write(\"Finished processing BAM file at \",datetime.now(),\". Total time taken \",total_time)\n",
    "\n",
    "bamFile.close()\n",
    "\n",
    "for barcode in barcode_dict.keys():\n",
    "    barcode_dict[barcode].computeHasEnoughCounts()\n",
    "\n",
    "write(len(CellBarcode.CellBarcodesWithEnoughCounts),\"Cell barcodes have enough UMIs for further processing.\")\n",
    "\n",
    "#Gives total number of elements that need to be computed. \n",
    "#Should be helpful when deciding how to divide the computation.\n",
    "def getTotalElementsInUpperTriangle(vector_length):\n",
    "    return int((vector_length*vector_length-vector_length)/2)\n",
    "\n",
    "#Dividing Upper Triangle into roughly equal parts for the Different Processes in Parallel\n",
    "def getEqualChunksForUpperTriangle(barcodes_list,max_cores):\n",
    "    vector_length=len(barcodes_list)\n",
    "    cumsum=0\n",
    "    chunk=getTotalElementsInUpperTriangle(vector_length)/max_cores\n",
    "    starting_i=[0]\n",
    "    max_val=chunk\n",
    "    ending_i=[]\n",
    "    for i in range(0,vector_length-1):\n",
    "        cumsum+=vector_length-1-i\n",
    "        if cumsum>max_val:\n",
    "            max_val=cumsum+chunk\n",
    "            starting_i.append(i)\n",
    "            ending_i.append(i-1)\n",
    "    ending_i.append(vector_length-1)\n",
    "    num_cores=len(starting_i)\n",
    "    output_pairs=[]\n",
    "    for position in range(0,len(starting_i)):\n",
    "        output_pairs.append((starting_i[position],ending_i[position]))\n",
    "    return output_pairs\n",
    "\n",
    "chunks=getEqualChunksForUpperTriangle(CellBarcode.CellBarcodesWithEnoughCounts,max_cores)\n",
    "num_cores=len(chunks)\n",
    "write(\"Dividing into the Following Chunks for Parallel Processing:\")\n",
    "write(chunks)\n",
    "write(\"Number of Cores that will be used: \",num_cores)\n",
    "\n",
    "pairs_list=[]\n",
    "\n",
    "chunk_lists=[]\n",
    "\n",
    "write(\"Collapsing Barcodes at a Hamming Distance of 1 and with the Major Barcode present above frequency\",SUBSTITUTION_ERROR_FREQ,\" at\",datetime.now())\n",
    "\n",
    "#Calculating Distances for the upper triangle of the distance matrix.\n",
    "def computeChunk(start,end,barcode_list):\n",
    "    def hammingDistance(barcode1,barcode2):\n",
    "        distance=0\n",
    "        i=0\n",
    "        while distance<2 and i<len(barcode1):\n",
    "            if barcode1[i]!=barcode2[i]:\n",
    "                distance+=1\n",
    "            i+=1\n",
    "        return distance\n",
    "\n",
    "    pairs_list=[]\n",
    "    vector_length=len(barcode_list)\n",
    "    for i in range(start,end+1):\n",
    "        for j in range(i+1,vector_length):\n",
    "            firstBarcode=barcode_list[i]\n",
    "            secondBarcode=barcode_list[j]\n",
    "            if hammingDistance(firstBarcode,secondBarcode)==1:\n",
    "                pairs_list.append((i,j))\n",
    "    return pairs_list\n",
    "\n",
    "with Pool(num_cores) as pool:\n",
    "    params=[]\n",
    "    for start,end in chunks:\n",
    "        barcodes_list_copy=copy.deepcopy(CellBarcode.CellBarcodesWithEnoughCounts)\n",
    "        params.append((start,end,barcodes_list_copy))\n",
    "    results=[pool.apply_async(computeChunk,p) for p in params]\n",
    "    for r in results:\n",
    "        chunk_lists.append(r.get())\n",
    "\n",
    "for sub_list in chunk_lists:\n",
    "    for pair in sub_list:\n",
    "        pairs_list.append(pair)\n",
    "\n",
    "for i,j in pairs_list:\n",
    "    firstBarcode=CellBarcode.CellBarcodesWithEnoughCounts[i]\n",
    "    secondBarcode=CellBarcode.CellBarcodesWithEnoughCounts[j]\n",
    "    barcode_dict[firstBarcode].combineIfSubstitution(barcode_dict[secondBarcode])\n",
    "    \n",
    "write(CellBarcode.NumberofBarcodesCollapsed,\"barcodes pairs were collapsed! At \",datetime.now())\n",
    "\n",
    "#Writing to new File with Updated Barcodes\n",
    "bamFile=pysam.AlignmentFile(bamFileName,\"rb\")\n",
    "BamRecords=bamFile.fetch(until_eof=True)\n",
    "outBamFile=pysam.AlignmentFile(outBamFileName,\"wb\",template=bamFile)\n",
    "\n",
    "start_time=datetime.now()\n",
    "prevMil=start_time\n",
    "write(\"Started Writing Updated BAM file at\",start_time,\".\")\n",
    "\n",
    "total_records=0\n",
    "for record in BamRecords:    \n",
    "\n",
    "    #For printing progress\n",
    "    total_records+=1\n",
    "    if total_records%1000000==0:\n",
    "        time_taken=datetime.now()-prevMil\n",
    "        write(\"Finished processing \",total_records,\"\\trecords at\",datetime.now(),\". Previous 1000000 Records took \",time_taken.total_seconds(),\"s\")\n",
    "        prevMil=datetime.now()\n",
    "    \n",
    "    #Main Logic\n",
    "    cell_barcode=record.get_tag('XC')\n",
    "    updated_barcode=barcode_dict[cell_barcode].barcodeToReturn.barcode #Getting Updated Barcode\n",
    "    record.set_tag('XC',updated_barcode)\n",
    "    outBamFile.write(record)\n",
    "\n",
    "total_time=datetime.now()-start_time\n",
    "write(\"Finished processing BAM file at \",datetime.now(),\". Total time taken \",total_time)\n",
    "\n",
    "bamFile.close()\n",
    "outBamFile.close()\n",
    "\n",
    "write(\"Total Execution Time:\",datetime.now()-script_start_time)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
