{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Author Kanishk Asthana kasthana@eng.ucsd.edu\n",
    "import pysam\n",
    "from datetime import datetime,timedelta\n",
    "import numpy as np\n",
    "import argparse\n",
    "import os\n",
    "import pygtrie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Default Values\n",
    "bamFileName=\"/home/kanishk/mixedHumanMouse/star_gene_exon_tagged_Long.sub_corrected.sample.sam\"\n",
    "outBamFileName=\"/home/kanishk/mixedHumanMouse/star_gene_exon_tagged_Long.sub_corrected.sample.syn_corrected.sam\"\n",
    "MIN_UMI_COUNT=20 #Minimum UMI count for Barcode to be considered for Collapse. Not counting UMIs for simplicity.\n",
    "num_cell_bases_missing=3 #Maximum number of Base synthesis errors from the end that will be checked for.\n",
    "cell_barcode_length=12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_file(input_filename):\n",
    "    if not os.path.isfile(input_filename):\n",
    "        raise argparse.ArgumentTypeError(\"File does not exist. Please use a valid file path.\")\n",
    "    return(input_filename)\n",
    "    \n",
    "def check_umi(input_umi):\n",
    "    value=int(input_umi)\n",
    "    assert value>0,\"Please enter positive values only for UMI Count Filter!\"\n",
    "    return value\n",
    "\n",
    "parser=argparse.ArgumentParser(description=\"Script to detect and Correct Synthesis errors in the Cell Barcodes. Please run CorrectSubstitutionErrors.py before running this script.\")\n",
    "parser.add_argument(\"INPUT_FILENAME\",help=\"Aligned Merged and Gene labeled BAM file for Correcting Synthesis Errors.\", type=parse_file)\n",
    "parser.add_argument(\"OUTPUT_FILENAME\",help=\"Please enter a Valid Path for the Error Corrected output BAM file.\",type=str)\n",
    "parser.add_argument(\"-umi\",\"--MIN_UMI\", help=\"Minimum UMI count per barcode to be considered for Collapse. Default Value is 20.\",type=check_umi)\n",
    "args=parser.parse_args()\n",
    "print(args)\n",
    "\n",
    "if args.MIN_UMI is not None:\n",
    "    MIN_UMI_COUNT=args.MIN_UMI\n",
    "\n",
    "bamFileName=args.INPUT_FILENAME\n",
    "outBamFileName=args.OUTPUT_FILENAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "script_start_time=datetime.now()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CellBarcode:\n",
    "    CellBarcodesWithEnoughCounts=[]\n",
    "    NumberofBarcodesCollapsed=0\n",
    "    def __init__(self,barcode,umi):\n",
    "        self.count=1 #Number of reads with that Barcode. When initializing you count the barcode you initialize with\n",
    "        self.barcode=barcode\n",
    "        self.wasCollapsed=False #This will be updated to True if this barcode gets collapsed.\n",
    "        self.umi_dict={}\n",
    "        self.umi_dict[umi]=1\n",
    "        self.umi_shift=0 #This number will track number of bases to shift the UMI by when doing the correction \n",
    "        self.num_unique_umi=1\n",
    "    \n",
    "    #For Debugging    \n",
    "    def show_info(self):\n",
    "        print(\"Current Value:\",self.barcode)\n",
    "        print(\"Was Collapsed?\",self.wasCollapsed)\n",
    "        print(\"UMI Shift Value:\",self.umi_shift)\n",
    "                \n",
    "    def increase_count(self,umi):\n",
    "        self.count+=1 #Increase count if you see a barcode\n",
    "        if umi in self.umi_dict:\n",
    "            self.umi_dict[umi]+=1\n",
    "        else:\n",
    "            self.umi_dict[umi]=1\n",
    "            \n",
    "    def computeHasEnoughCounts(self):\n",
    "        self.num_unique_umi=len(self.umi_dict.keys())\n",
    "        if self.num_unique_umi>MIN_UMI_COUNT:\n",
    "            CellBarcode.CellBarcodesWithEnoughCounts.append(self.barcode)\n",
    "            \n",
    "    def collapseBarcode(self,new_barcode):\n",
    "        self.wasCollapsed=True\n",
    "        self.umi_shift+=1\n",
    "        self.barcode=new_barcode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_dict={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started Processing BAM file at 2021-01-28 19:48:18.417991 . Getting Cell Barcodes to correct Illumina Sequencing Base Substitution Errors!\n",
      "Finished processing BAM file at  2021-01-28 19:48:20.744544 . Total time taken  0:00:02.326467\n",
      "2197 Cell barcodes have enough UMIs for further processing.\n"
     ]
    }
   ],
   "source": [
    "#If you get an error here your file probably not correctly formated. Make sure you have a header.\n",
    "bamFile=pysam.AlignmentFile(bamFileName,\"rb\")\n",
    "BamRecords=bamFile.fetch(until_eof=True)\n",
    "\n",
    "start_time=datetime.now()\n",
    "prevMil=start_time\n",
    "print(\"Started Processing BAM file at\",start_time,\". Getting Cell Barcodes to correct Illumina Sequencing Base Substitution Errors!\")\n",
    "\n",
    "total_records=0\n",
    "for record in BamRecords:\n",
    "    \n",
    "\n",
    "    #For printing progress\n",
    "    total_records+=1\n",
    "    if total_records%1000000==0:\n",
    "        time_taken=datetime.now()-prevMil\n",
    "        print(\"Finished processing \",total_records,\"\\trecords at\",datetime.now(),\". Previous 1000000 Records took \",time_taken.total_seconds(),\"s\")\n",
    "        prevMil=datetime.now()\n",
    "    \n",
    "\n",
    "    #Main Logic\n",
    "    cell_barcode=record.get_tag('XC')\n",
    "    umi=record.get_tag('XM')\n",
    "    \n",
    "    if cell_barcode in barcode_dict:\n",
    "        barcode_dict[cell_barcode].increase_count(umi)\n",
    "    else:\n",
    "        barcode_dict[cell_barcode]=CellBarcode(cell_barcode,umi)\n",
    "\n",
    "\n",
    "total_time=datetime.now()-start_time\n",
    "print(\"Finished processing BAM file at \",datetime.now(),\". Total time taken \",total_time)\n",
    "\n",
    "bamFile.close()\n",
    "\n",
    "for barcode in barcode_dict.keys():\n",
    "    barcode_dict[barcode].computeHasEnoughCounts()\n",
    "\n",
    "print(len(CellBarcode.CellBarcodesWithEnoughCounts),\"Cell barcodes have enough UMIs for further processing.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Building Barcode Trie:\n",
    "barcode_trie=pygtrie.CharTrie()\n",
    "for barcode in CellBarcode.CellBarcodesWithEnoughCounts:\n",
    "    barcode_trie[barcode]=barcode_dict[barcode]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Recursively Update all objects in Nested Lists made in Barcode_Trie\n",
    "def update(new_barcode,obj_list):\n",
    "    print(\"For new barcode:\",new_barcode)\n",
    "    for obj in obj_list:\n",
    "        if type(obj) is CellBarcode:\n",
    "            print(obj.barcode,\"has\",obj.num_unique_umi,\"UMIS\")\n",
    "            obj.collapseBarcode(new_barcode)\n",
    "        else:\n",
    "            update(new_barcode,obj)\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking for Cell Barcode Synthesis Errors..\n",
      "Checking for Synthesis Errors in Cell Barcode at Base Position  12\n",
      "CATAGTCAAAT  has  4  barcodes associated with it.\n",
      "ATCTAACCTCA  has  4  barcodes associated with it.\n",
      "ATAGAACATCG  has  4  barcodes associated with it.\n",
      "ATAGAACATCC  has  4  barcodes associated with it.\n",
      "AAGGTCCTACC  has  4  barcodes associated with it.\n",
      "ATAGAACATCA  has  4  barcodes associated with it.\n",
      "ATAGAACATCT  has  4  barcodes associated with it.\n",
      "CCAGCTGGCGG  has  4  barcodes associated with it.\n",
      "GTCTAGGTCCA  has  4  barcodes associated with it.\n",
      "GTCTAGGTCCT  has  4  barcodes associated with it.\n",
      "ACTCTTCGTAC  has  4  barcodes associated with it.\n",
      "GCTTGTTGTTG  has  4  barcodes associated with it.\n",
      "AATCCCGCATA  has  4  barcodes associated with it.\n",
      "TAGTCTCCTAT  has  4  barcodes associated with it.\n",
      "ATCGACACCCG  has  4  barcodes associated with it.\n",
      "GTCTAGGTCCC  has  4  barcodes associated with it.\n",
      "AATCCCGCATG  has  4  barcodes associated with it.\n",
      "GTGTCCGCCTG  has  4  barcodes associated with it.\n",
      "AATCCCGCATC  has  4  barcodes associated with it.\n",
      "GTCTAGGTCCG  has  4  barcodes associated with it.\n",
      "GATTCACCAAT  has  4  barcodes associated with it.\n",
      "CGGCTAATTTG  has  4  barcodes associated with it.\n",
      "AATCCCGCATT  has  4  barcodes associated with it.\n",
      "ATCCCTTAAGT  has  4  barcodes associated with it.\n",
      "ATCCTTGACAC  has  4  barcodes associated with it.\n",
      "25 subtries found at this position.\n",
      "For new barcode: CATAGTCAAATN\n",
      "CATAGTCAAATT has 434 UMIS\n",
      "CATAGTCAAATG has 598 UMIS\n",
      "CATAGTCAAATA has 401 UMIS\n",
      "CATAGTCAAATC has 413 UMIS\n",
      "\n",
      "For new barcode: ATCTAACCTCAN\n",
      "ATCTAACCTCAG has 1415 UMIS\n",
      "ATCTAACCTCAA has 1153 UMIS\n",
      "ATCTAACCTCAC has 861 UMIS\n",
      "ATCTAACCTCAT has 1205 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCGN\n",
      "ATAGAACATCGG has 1317 UMIS\n",
      "ATAGAACATCGC has 923 UMIS\n",
      "ATAGAACATCGA has 1098 UMIS\n",
      "ATAGAACATCGT has 1068 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCCN\n",
      "ATAGAACATCCT has 836 UMIS\n",
      "ATAGAACATCCC has 704 UMIS\n",
      "ATAGAACATCCG has 838 UMIS\n",
      "ATAGAACATCCA has 749 UMIS\n",
      "\n",
      "For new barcode: AAGGTCCTACCN\n",
      "AAGGTCCTACCA has 1335 UMIS\n",
      "AAGGTCCTACCC has 1354 UMIS\n",
      "AAGGTCCTACCG has 1359 UMIS\n",
      "AAGGTCCTACCT has 1413 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCAN\n",
      "ATAGAACATCAG has 1057 UMIS\n",
      "ATAGAACATCAT has 970 UMIS\n",
      "ATAGAACATCAC has 831 UMIS\n",
      "ATAGAACATCAA has 798 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCTN\n",
      "ATAGAACATCTC has 860 UMIS\n",
      "ATAGAACATCTT has 165 UMIS\n",
      "ATAGAACATCTG has 94 UMIS\n",
      "ATAGAACATCTA has 94 UMIS\n",
      "\n",
      "For new barcode: CCAGCTGGCGGN\n",
      "CCAGCTGGCGGT has 180 UMIS\n",
      "CCAGCTGGCGGG has 239 UMIS\n",
      "CCAGCTGGCGGA has 229 UMIS\n",
      "CCAGCTGGCGGC has 198 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCAN\n",
      "GTCTAGGTCCAC has 113 UMIS\n",
      "GTCTAGGTCCAT has 118 UMIS\n",
      "GTCTAGGTCCAA has 115 UMIS\n",
      "GTCTAGGTCCAG has 126 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCTN\n",
      "GTCTAGGTCCTA has 98 UMIS\n",
      "GTCTAGGTCCTG has 129 UMIS\n",
      "GTCTAGGTCCTC has 83 UMIS\n",
      "GTCTAGGTCCTT has 110 UMIS\n",
      "\n",
      "For new barcode: ACTCTTCGTACN\n",
      "ACTCTTCGTACA has 196 UMIS\n",
      "ACTCTTCGTACG has 248 UMIS\n",
      "ACTCTTCGTACT has 181 UMIS\n",
      "ACTCTTCGTACC has 161 UMIS\n",
      "\n",
      "For new barcode: GCTTGTTGTTGN\n",
      "GCTTGTTGTTGA has 46 UMIS\n",
      "GCTTGTTGTTGG has 72 UMIS\n",
      "GCTTGTTGTTGC has 29 UMIS\n",
      "GCTTGTTGTTGT has 31 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATAN\n",
      "AATCCCGCATAA has 21 UMIS\n",
      "AATCCCGCATAT has 33 UMIS\n",
      "AATCCCGCATAG has 32 UMIS\n",
      "AATCCCGCATAC has 31 UMIS\n",
      "\n",
      "For new barcode: TAGTCTCCTATN\n",
      "TAGTCTCCTATG has 159 UMIS\n",
      "TAGTCTCCTATT has 107 UMIS\n",
      "TAGTCTCCTATC has 75 UMIS\n",
      "TAGTCTCCTATA has 78 UMIS\n",
      "\n",
      "For new barcode: ATCGACACCCGN\n",
      "ATCGACACCCGT has 36 UMIS\n",
      "ATCGACACCCGG has 55 UMIS\n",
      "ATCGACACCCGA has 43 UMIS\n",
      "ATCGACACCCGC has 47 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCCN\n",
      "GTCTAGGTCCCG has 157 UMIS\n",
      "GTCTAGGTCCCA has 97 UMIS\n",
      "GTCTAGGTCCCT has 133 UMIS\n",
      "GTCTAGGTCCCC has 88 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATGN\n",
      "AATCCCGCATGG has 72 UMIS\n",
      "AATCCCGCATGA has 34 UMIS\n",
      "AATCCCGCATGT has 48 UMIS\n",
      "AATCCCGCATGC has 43 UMIS\n",
      "\n",
      "For new barcode: GTGTCCGCCTGN\n",
      "GTGTCCGCCTGA has 46 UMIS\n",
      "GTGTCCGCCTGC has 37 UMIS\n",
      "GTGTCCGCCTGT has 31 UMIS\n",
      "GTGTCCGCCTGG has 69 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATCN\n",
      "AATCCCGCATCG has 44 UMIS\n",
      "AATCCCGCATCA has 36 UMIS\n",
      "AATCCCGCATCT has 45 UMIS\n",
      "AATCCCGCATCC has 31 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCGN\n",
      "GTCTAGGTCCGC has 127 UMIS\n",
      "GTCTAGGTCCGT has 111 UMIS\n",
      "GTCTAGGTCCGG has 202 UMIS\n",
      "GTCTAGGTCCGA has 108 UMIS\n",
      "\n",
      "For new barcode: GATTCACCAATN\n",
      "GATTCACCAATA has 48 UMIS\n",
      "GATTCACCAATG has 49 UMIS\n",
      "GATTCACCAATT has 55 UMIS\n",
      "GATTCACCAATC has 44 UMIS\n",
      "\n",
      "For new barcode: CGGCTAATTTGN\n",
      "CGGCTAATTTGT has 30 UMIS\n",
      "CGGCTAATTTGC has 21 UMIS\n",
      "CGGCTAATTTGG has 33 UMIS\n",
      "CGGCTAATTTGA has 31 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATTN\n",
      "AATCCCGCATTA has 38 UMIS\n",
      "AATCCCGCATTG has 51 UMIS\n",
      "AATCCCGCATTC has 47 UMIS\n",
      "AATCCCGCATTT has 53 UMIS\n",
      "\n",
      "For new barcode: ATCCCTTAAGTN\n",
      "ATCCCTTAAGTA has 24 UMIS\n",
      "ATCCCTTAAGTG has 26 UMIS\n",
      "ATCCCTTAAGTT has 34 UMIS\n",
      "ATCCCTTAAGTC has 24 UMIS\n",
      "\n",
      "For new barcode: ATCCTTGACACN\n",
      "ATCCTTGACACA has 78 UMIS\n",
      "ATCCTTGACACC has 89 UMIS\n",
      "ATCCTTGACACT has 82 UMIS\n",
      "ATCCTTGACACG has 75 UMIS\n",
      "\n",
      "Checking for Synthesis Errors in Cell Barcode at Base Position  11\n",
      "ATAGAACATC  has  4  barcodes associated with it.\n",
      "GTCTAGGTCC  has  4  barcodes associated with it.\n",
      "AATCCCGCAT  has  4  barcodes associated with it.\n",
      "3 subtries found at this position.\n",
      "For new barcode: ATAGAACATCNN\n",
      "For new barcode: ATAGAACATCNN\n",
      "ATAGAACATCGN has 1317 UMIS\n",
      "ATAGAACATCGN has 923 UMIS\n",
      "ATAGAACATCGN has 1098 UMIS\n",
      "ATAGAACATCGN has 1068 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCNN\n",
      "ATAGAACATCCN has 836 UMIS\n",
      "ATAGAACATCCN has 704 UMIS\n",
      "ATAGAACATCCN has 838 UMIS\n",
      "ATAGAACATCCN has 749 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCNN\n",
      "ATAGAACATCAN has 1057 UMIS\n",
      "ATAGAACATCAN has 970 UMIS\n",
      "ATAGAACATCAN has 831 UMIS\n",
      "ATAGAACATCAN has 798 UMIS\n",
      "\n",
      "For new barcode: ATAGAACATCNN\n",
      "ATAGAACATCTN has 860 UMIS\n",
      "ATAGAACATCTN has 165 UMIS\n",
      "ATAGAACATCTN has 94 UMIS\n",
      "ATAGAACATCTN has 94 UMIS\n",
      "\n",
      "\n",
      "For new barcode: GTCTAGGTCCNN\n",
      "For new barcode: GTCTAGGTCCNN\n",
      "GTCTAGGTCCAN has 113 UMIS\n",
      "GTCTAGGTCCAN has 118 UMIS\n",
      "GTCTAGGTCCAN has 115 UMIS\n",
      "GTCTAGGTCCAN has 126 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCNN\n",
      "GTCTAGGTCCTN has 98 UMIS\n",
      "GTCTAGGTCCTN has 129 UMIS\n",
      "GTCTAGGTCCTN has 83 UMIS\n",
      "GTCTAGGTCCTN has 110 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCNN\n",
      "GTCTAGGTCCCN has 157 UMIS\n",
      "GTCTAGGTCCCN has 97 UMIS\n",
      "GTCTAGGTCCCN has 133 UMIS\n",
      "GTCTAGGTCCCN has 88 UMIS\n",
      "\n",
      "For new barcode: GTCTAGGTCCNN\n",
      "GTCTAGGTCCGN has 127 UMIS\n",
      "GTCTAGGTCCGN has 111 UMIS\n",
      "GTCTAGGTCCGN has 202 UMIS\n",
      "GTCTAGGTCCGN has 108 UMIS\n",
      "\n",
      "\n",
      "For new barcode: AATCCCGCATNN\n",
      "For new barcode: AATCCCGCATNN\n",
      "AATCCCGCATAN has 21 UMIS\n",
      "AATCCCGCATAN has 33 UMIS\n",
      "AATCCCGCATAN has 32 UMIS\n",
      "AATCCCGCATAN has 31 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATNN\n",
      "AATCCCGCATGN has 72 UMIS\n",
      "AATCCCGCATGN has 34 UMIS\n",
      "AATCCCGCATGN has 48 UMIS\n",
      "AATCCCGCATGN has 43 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATNN\n",
      "AATCCCGCATCN has 44 UMIS\n",
      "AATCCCGCATCN has 36 UMIS\n",
      "AATCCCGCATCN has 45 UMIS\n",
      "AATCCCGCATCN has 31 UMIS\n",
      "\n",
      "For new barcode: AATCCCGCATNN\n",
      "AATCCCGCATTN has 38 UMIS\n",
      "AATCCCGCATTN has 51 UMIS\n",
      "AATCCCGCATTN has 47 UMIS\n",
      "AATCCCGCATTN has 53 UMIS\n",
      "\n",
      "\n",
      "Checking for Synthesis Errors in Cell Barcode at Base Position  10\n",
      "0 subtries found at this position.\n"
     ]
    }
   ],
   "source": [
    "print(\"Checking for Cell Barcode Synthesis Errors..\")\n",
    "for base_position in range(cell_barcode_length-1,cell_barcode_length-1-num_cell_bases_missing,-1):\n",
    "    print(\"Checking for Synthesis Errors in Cell Barcode at Base Position \",base_position+1)\n",
    "    barcodes_to_be_combined=[]\n",
    "    current_barcodes=CellBarcode.CellBarcodesWithEnoughCounts\n",
    "    \n",
    "    #Looking for Barcodes with 4 neighbours present at base position\n",
    "    for cell_barcode in current_barcodes:\n",
    "        prefix=cell_barcode[0:base_position]\n",
    "        try:\n",
    "            #If 4 values in the list are associated with that prefix then pop the node and store it in a list\n",
    "            subtrie_keys=barcode_trie.keys(prefix)\n",
    "            subtrie=barcode_trie.items(prefix)\n",
    "            if len(subtrie_keys)==4:\n",
    "                print(prefix,\" has \",len(subtrie_keys),\" barcodes associated with it.\")\n",
    "                barcodes_to_be_combined.append(subtrie)\n",
    "                del barcode_trie[prefix:] #Popping Barcodes in Subtrie from Main Trie\n",
    "        except KeyError: #Catching a key error if barcode already popped from the main Trie\n",
    "            continue\n",
    "    print(len(barcodes_to_be_combined),\"subtries found at this position.\")\n",
    "    \n",
    "    #Collapsing Barcodes\n",
    "    for subtrie in barcodes_to_be_combined:\n",
    "        barcode,obj_to_update=subtrie[0] #Getting First element for Generating new Barcode\n",
    "        new_barcode=list(barcode)\n",
    "        new_barcode[base_position]=\"N\"#Replacing Base with N\n",
    "        new_barcode=\"\".join(new_barcode) #Converting back to String\n",
    "        obj_list=[]\n",
    "        for barcode,obj in subtrie:\n",
    "            obj_list.append(obj)\n",
    "        update(new_barcode,obj_list) #Recursively update all elements of obj_list\n",
    "        barcode_trie[new_barcode]=obj_list #Adding combined Tag back to the Trie as Nested List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 barcodes were collapsed!\n"
     ]
    }
   ],
   "source": [
    "#Printing total number of barcodes collapsed\n",
    "collapse_count=0\n",
    "for barcode in barcode_dict.keys():\n",
    "    if barcode_dict[barcode].wasCollapsed:\n",
    "        collapse_count+=1\n",
    "print(collapse_count,\"barcodes were collapsed!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getNewUMI(original_cell_barcode,updated_barcode_object,original_umi):\n",
    "    if updated_barcode_object.wasCollapsed:\n",
    "        cell_barcode_tail=original_cell_barcode[-updated_barcode_object.umi_shift:]\n",
    "        umi_head=original_umi[:-updated_barcode_object.umi_shift]\n",
    "        newUMI=cell_barcode_tail+umi_head\n",
    "        return newUMI\n",
    "    else:\n",
    "        return original_umi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started Writing Updated BAM file at 2021-01-28 19:48:21.201478 .\n",
      "Finished processing BAM file at  2021-01-28 19:48:24.904487 . Total time taken  0:00:03.702904\n"
     ]
    }
   ],
   "source": [
    "#Writing to new File with Updated Barcodes\n",
    "bamFile=pysam.AlignmentFile(bamFileName,\"rb\")\n",
    "BamRecords=bamFile.fetch(until_eof=True)\n",
    "outBamFile=pysam.AlignmentFile(outBamFileName,\"w\",template=bamFile)\n",
    "\n",
    "start_time=datetime.now()\n",
    "prevMil=start_time\n",
    "print(\"Started Writing Updated BAM file at\",start_time,\".\")\n",
    "\n",
    "total_records=0\n",
    "for record in BamRecords:    \n",
    "\n",
    "    #For printing progress\n",
    "    total_records+=1\n",
    "    if total_records%1000000==0:\n",
    "        time_taken=datetime.now()-prevMil\n",
    "        print(\"Finished processing \",total_records,\"\\trecords at\",datetime.now(),\". Previous 1000000 Records took \",time_taken.total_seconds(),\"s\")\n",
    "        prevMil=datetime.now()\n",
    "    \n",
    "    #Main Logic\n",
    "    original_cell_barcode=record.get_tag('XC')\n",
    "    original_umi=record.get_tag(\"XM\")\n",
    "    updated_barcode=barcode_dict[original_cell_barcode] #Getting Updated Barcode for Query Name\n",
    "    new_umi=getNewUMI(original_cell_barcode,updated_barcode,original_umi)#Will return updated if changed, original if unchanged.\n",
    "    record.set_tag('XC',updated_barcode.barcode)\n",
    "    record.set_tag('XM',new_umi)\n",
    "    outBamFile.write(record)\n",
    "\n",
    "total_time=datetime.now()-start_time\n",
    "print(\"Finished processing BAM file at \",datetime.now(),\". Total time taken \",total_time)\n",
    "\n",
    "bamFile.close()\n",
    "outBamFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Execution Time: 0:00:06.663374\n"
     ]
    }
   ],
   "source": [
    "print(\"Total Execution Time:\",datetime.now()-script_start_time)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
